src/dos/dos.cpp

```diff
@@ -44,7 +44,30 @@ const Bit8u DOS_DATE_months[] = {
 	0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31
 };
 
-static void DOS_AddDays(Bitu days) {
+uint16_t DOS_PackTime(uint16_t hour, uint16_t min, uint16_t sec)
+{
+	const auto hour_portion = (hour & 0x1f) << 11;
+	const auto min_portion = (min & 0x3f) << 5;
+	const auto sec_portion = (sec / 2) & 0x1f;
+	const auto packed = hour_portion | min_portion | sec_portion;
+
+	assert(packed >= 0 && packed <= UINT16_MAX); // ensure rvalue won't wrap
+	return static_cast<uint16_t>(packed);
+}
+
+uint16_t DOS_PackDate(uint16_t year, uint16_t mon, uint16_t day)
+{
+	const auto year_portion = ((year - 1980) & 0x7f) << 9;
+	const auto mon_portion = (mon & 0x3f) << 5;
+	const auto day_portion = day & 0x1f;
+	const auto packed = year_portion | mon_portion | day_portion;
+
+	assert(packed >= 0 && packed <= UINT16_MAX); // ensure rvalue won't wrap
+	return static_cast<uint16_t>(packed);
+}
```

Comment:
Very nice - it's the same as the previous version, but more readable :)

I was considering writing the same this way:
```C++
	const auto h_bits = 0b1111100000000000 & (hour << 11);
	const auto m_bits = 0b0000011111100000 & (min << 5);
	const auto s_bits = 0b0000000000011111 & (sec / 2);
	const auto packed = h_bits | m_bits | s_bits;
```
So maybe it's obvious enough, that even `assert` would not be needed?

Reply:
@dreamer, this is a real treat! 

You've improved the readability to instructional levels, to where each conversion is now easily understood including flattening the three adjacent  5-bit-long segments into the final non-overlapping 15-bit result.

(And given it's now completely obvious the result is 15-bits starting at the LSB, there's no need to `assert` the resulting bounds like you mentioned.)

Thinking about this improved representation (ie: the GUS could benefit from it greatly), is there a rote way we could perform this translation? (ie: also shifting the `and`-value so we can re-order the operations?).  I'm thinking maybe some Python that performs the translation and even tests the before and after with a range of values.

Reply:
The resulting value is 16 bits ;) (storing 0-59) minutes requires 6 bits. I was surprised to notice it happens at the expense of seconds resolution (I forgot DOS uses 2s resolution for timestamps).

> Thinking about this improved representation (ie: the GUS could benefit from it greatly), is there a rote way we could perform this translation? (ie: also shifting the and-value so we can re-order the operations?). I'm thinking maybe some Python that performs the translation and even tests the before and after with a range of values.

There's [`spatch` program (semantic patch)](https://coccinelle.gitlabpages.inria.fr/website/sp.html) for creating such automatic refactors (it's often used by Linux kernel devs AFAIK).

Reply:
But now I'm worried about PackDate function… is seems like old implementation had a tiny problem with using wrong mask for months? 0x3f is 7 bits, when I guess months should be stored on 4 bits?

I wonder if this bug exists intentionally to reproduce some MS-DOS bug?

Reply:
MSDN does not refer to any bugs in old DOS time format... https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-filetimetodosdatetime months should be stored on 4 bits.

Reply:
Thanks for the spatch reference; I haven't heard of this tool -- very interesting!
Regarding the month - do you want to correct it?

Reply:
Maybe let's fix it during the review :)
```C++
	const auto y_bits = 0b1111111000000000 & ((year - 1980) << 9);
	const auto m_bits = 0b0000000111100000 & (mon << 5);
	const auto d_bits = 0b0000000000011111 & day;
	const auto packed = y_bits | m_bits | d_bits;
```

Reply:
Dropped it in; and switched you to authoring the commit.
Thanks for the very nice implementation @dreamer :-)